### DISCUSSION TITLE
Docker 컨테이너 운영 방식

### DISCUSSION THREAD

**Discussion by hooniegit** [hanul-pipeline/stopped_hanul-site-pipeline#12] [Q&A] [answered]
# 들어가며
(많은 에너지를 쏟는 시간이 아닙니다! 느긋하게 쉬다가 여유가 생기는 때에 함께 의논해 봅시다 :) )

### 🧑‍🏫 개요
Airflow 관련 기능을 탑제하기 전인 현재, 그 동안 우리가 추구했던 컨테이너의 운영 방식은 아래와 같았습니다.

1. Dockerfile 내부에 필요로 하는 모든 환경 설정을 입력
2. docker-compose.yml 내부에 포트 개방 설정 등을 입력
=> 단일 컨테이너에 필요로 하는 모든 기능들을 설치

그런데 현재 airflow - python - mysql-connector-python 사이의 버전 호환 문제가 발생하면서 하나의 Dockerfile에 모든 정보들을 입력하는 방식에 다소 무리가 있다는 것을 확인하게 되었습니다. 테스트 수행에 소모되는 시간이 컨테이너 빌드 속도에 비례함에 따라 기하급수적으로 증가하고 있으며, 버전 호환성을 고려해야 하는 툴들이 많아짐에 따라 이러한 문제점이 크게 부각되는 중입니다.

### 🙏 주제 및 아이디어 브랜칭
현재 고수하고 있던 도커 컨테이너 운영 방식에 다소 어려움이 있을 것임을 고려하여, Dockerfile 및 docker-compose.yml 파일의 작성 및 컨테이너 운영 방식 및 방향성에 대한 약간의 수정이 필요할 것 같습니다.
어떻게 하면 기존 방식이 가지는 한계점들을 개선할 수 있을까요?
간단한 답변이라도 좋습니다. 편하게 작성 부탁드립니다!

**Comment by blahxblah** [comment, marked as answer]
기존의 도커파일에 airflow를 추가하는데 호환성 문제가 계속 발생하고 airflow의 자동화 작업이 계속 유지되어야 하는 상황에서 airflow만의 도커 컨테이너를 따로 작업하여 파이프라인의 자동화 안정성을 높이는게 좋은 것 같습니다.
그래서 도커 컨테이너를 2개를 운용하면서 각각 따로 서버에 설치하여 workflow에 문제가 없도록 하는게 나을 것 같아요.

**Reply by hooniegit** [reply]
오.. 그러면 docker-compose.yml 파일을 통해 
1. 현재 사용중인 ubuntu 기반의 도커 컨테이너
2. dockerhub의 airflow 이미지 기반의 도커 컨테이너
를 동시에 띄우고, 둘이 서로 통신 가능하도록 설정하여 컨테이너를 분산시킬 수도 있겠네요?

월요일 모임에서 조금 더 이야기를 나누면 좋을 것 같네요 :)
관련 자료를 조금 더 찾아보겠습니다, 감사합니다!

**Reply by hooniegit** [reply]
/+ 저희가 기존에 사용하고 있던 GCP 내에 별도의 컨테이너를 위치시켜 포트포워딩을 해 주어도 좋을 것 같네요 💡
